### ISSUE TITLE
npm in MINGW throws: Cannot find module 'C:\c\Program Files\nodejs...'

### ISSUE THREAD

**Issue by valoricDe** [npm/npm#18499]
npm Version: 3.10.10
    Platform: MINGW64_NT-6.1 2.5.0(0.295/5/3) 2016-03-31 18:47 x86_64 Msys

When I run npm (/c/Program Files/nodejs/npm) in a git bash (which uses MINGW) the npm script throws:
Error: Cannot find module 'C:\c\Program Files\nodejs\node_modules\npm\bin\npm-cli.js'

The problem is that the npm shell script executes a window node.exe but supplies it with a unix path for the npm-cli.js.

As I see it the fix is to also convert the path with cygpath for MINGW as well in /c/Program Files/nodejs/npm line 6-8:

```
case `uname` in
    *CYGWIN*) basedir=`cygpath -w "$basedir"`;;
    *MINGW*) basedir=`cygpath -w "$basedir"`;;
esac
```

Migrated from https://github.com/nodejs/node/issues/15356

**Comment by NullSoldier** [comment]
Running into this issue in a pretty recent install of bash for windows (msys)

**Comment by denniske** [comment]
I would also appreciate MSYS to be added:
```
case `uname` in
    *CYGWIN*) basedir=`cygpath -w "$basedir"`;;
    *MINGW*) basedir=`cygpath -w "$basedir"`;;
    *MSYS*) basedir=`cygpath -w "$basedir"`;;
esac
```
Or is there a workaround? Maybe telling MSYS to set uname to CYGWIN or something like that?

**Comment by bimlas** [comment]
I did some investigation on this issue and found an interesting solution.

After `npm install -g diff-so-fancy` (for example) this startup script is created in `%APPDATA%\npm`:

%APPDATA%\npm\diff-so-fancy
```sh
#!/bin/sh
basedir=$(dirname "$(echo "$0" | sed -e 's,\\,/,g')")

case `uname` in
    *CYGWIN*) basedir=`cygpath -w "$basedir"`;;
esac

if [ -x "$basedir/perl" ]; then
  "$basedir/perl"  "$basedir/node_modules/diff-so-fancy/diff-so-fancy" "$@"
  ret=$?
else 
  perl  "$basedir/node_modules/diff-so-fancy/diff-so-fancy" "$@"
  ret=$?
fi
exit $ret
```

The issue is about the first two lines (which is the same in every npm module startup scripts):

```sh
#!/bin/sh
basedir=$(dirname "$(echo "$0" | sed -e 's,\\,/,g')")
```

`basedir` is empty because of the shebang line. Tested on MSys Git Bash:

test/tmp.sh
```sh
#!/bin/sh
echo "A) $0"
echo "B) $0" | sed -e 's,\\,/,g'
echo "C) $(echo "$0" | sed -e 's,\\,/,g')"
echo "D) $(dirname "$(echo "$0" | sed -e 's,\\,/,g')")"
```

Output
```
$ test/tmp.sh
A) test/tmp.sh
B) test/tmp.sh
C)
D)
```

Changing `#!/bin/sh` to `#!/bin/dash`
```
$ test/tmp.sh
A) test/tmp.sh
B) test/tmp.sh
C) test/tmp.sh
D) test
```

Note that **`$(...)` is a Bash expression, not `sh`**! The funny part is that is not working with `#!/bin/bash`... (**EDIT** Because `/bin/bash` and `/bin/sh` are copies of each other)

![screencast](https://i.imgur.com/WIVutLp.gif)

**Comment by bimlas** [comment]
Found the issue on https://github.com/git-for-windows/git/issues/1619

**Comment by bimlas** [comment]
TL;DR: My antivirus (Comodo Internet Security) caused the problem. Turning it off does not solved the issue, but uninstalling did it.

**EDIT**

Just reported to Comodo as false-positive.

**Comment by bimlas** [comment]
Another solution is here: https://github.com/git-for-windows/git/issues/1562#issuecomment-378756798

**Comment by valoricDe** [comment]
Hey guys. I'm not developing on a windows machine and therefore can't support this issue anymore. Feel free to comment, but I will close this ticket now xD
